undertow/src/test/java/org/wildfly/elytron/web/undertow/server/util/InfinispanSessionManager.java

```diff
@@ -0,0 +1,611 @@
+ *
+ * @author <a href="mailto:psilva@redhat.com">Pedro Igor</a>
+ */
+public class InfinispanSessionManager implements SessionManager, SessionManagerStatistics {
+
+    private final AttachmentKey<SessionImpl> NEW_SESSION = AttachmentKey.create(SessionImpl.class);
+
+    private final SessionIdGenerator sessionIdGenerator;
+
+    private ConcurrentMap<String, SessionImpl> sessions;
+
+    private final SessionListeners sessionListeners = new SessionListeners();
+
+    /**
+     * 30 minute default
+     */
+    private volatile int defaultSessionTimeout = 30 * 60;
+
+    private final int maxSize;
+
+    private final ConcurrentDirectDeque<String> evictionQueue;
+
+    private final String deploymentName;
+
+    private final AtomicLong createdSessionCount = new AtomicLong();
+    private final AtomicLong expiredSessionCount = new AtomicLong();
+    private final AtomicLong rejectedSessionCount = new AtomicLong();
+    private final AtomicLong averageSessionLifetime = new AtomicLong();
+    private final AtomicLong longestSessionLifetime = new AtomicLong();
+    private final boolean statisticsEnabled;
+
+    private volatile long startTime;
+
+    private final boolean expireOldestUnusedSessionOnMax;
+    private EmbeddedCacheManager cacheManager;
+
+
+    public InfinispanSessionManager(String deploymentName, int maxSessions, boolean expireOldestUnusedSessionOnMax) {
+        this(new SecureRandomSessionIdGenerator(), deploymentName, maxSessions, expireOldestUnusedSessionOnMax);
+    }
+
+    public InfinispanSessionManager(SessionIdGenerator sessionIdGenerator, String deploymentName, int maxSessions, boolean expireOldestUnusedSessionOnMax) {
+        this(sessionIdGenerator, deploymentName, maxSessions, expireOldestUnusedSessionOnMax, true);
+    }
+
+    public InfinispanSessionManager(SessionIdGenerator sessionIdGenerator, String deploymentName, int maxSessions, boolean expireOldestUnusedSessionOnMax, boolean statisticsEnabled) {
+        this.sessionIdGenerator = sessionIdGenerator;
+        this.deploymentName = deploymentName;
+        this.statisticsEnabled = statisticsEnabled;
+        this.expireOldestUnusedSessionOnMax = expireOldestUnusedSessionOnMax;
+        this.maxSize = maxSessions;
+        ConcurrentDirectDeque<String> evictionQueue = null;
+        if (maxSessions > 0) {
+            evictionQueue = ConcurrentDirectDeque.newInstance();
+        }
+        this.evictionQueue = evictionQueue;
+    }
+
+    public InfinispanSessionManager(String deploymentName, int maxSessions) {
+        this(deploymentName, maxSessions, false);
+    }
+
+    public InfinispanSessionManager(String id) {
+        this(id, -1);
+    }
+
+    @Override
+    public String getDeploymentName() {
+        return this.deploymentName;
+    }
+
+    @Override
+    public void start() {
+        createdSessionCount.set(0);
+        expiredSessionCount.set(0);
+        this.cacheManager = new DefaultCacheManager(
```

Comment:
Contributors to the security related projects may not necessarily have experience of the Infinispan configuration - I think it would be nice to have some comments describing how you have configured the cache manager and why so for future updates we don't overlook decisions you may have made.